src/Elliptic/Systems/Xcts/Equations.cpp

```diff
@@ -0,0 +1,779 @@
+void fully_contract_flat_cartesian(
+    const gsl::not_null<Scalar<DataVector>*> result,
+    const tnsr::II<DataVector, 3>& tensor1,
+    const tnsr::II<DataVector, 3>& tensor2) noexcept {
+  get(*result) = tensor1.multiplicity(0_st) * tensor1[0] * tensor2[0];
+  for (size_t i = 1; i < tensor1.size(); ++i) {
+    get(*result) += tensor1.multiplicity(i) * tensor1[i] * tensor2[i];
+  }
+}
+void fully_contract(const gsl::not_null<Scalar<DataVector>*> result,
+                    const gsl::not_null<DataVector*> buffer1,
+                    const gsl::not_null<DataVector*> buffer2,
+                    const tnsr::II<DataVector, 3>& tensor1,
+                    const tnsr::II<DataVector, 3>& tensor2,
+                    const tnsr::ii<DataVector, 3>& metric) noexcept {
+  get(*result) = 0.;
+  for (size_t i = 0; i < 3; ++i) {
+    for (size_t j = 0; j < 3; ++j) {
+      *buffer1 = metric.get(i, 0) * tensor1.get(0, j);
+      *buffer2 = metric.get(j, 0) * tensor2.get(i, 0);
+      for (size_t k = 1; k < 3; ++k) {
+        *buffer1 += metric.get(i, k) * tensor1.get(k, j);
+        *buffer2 += metric.get(j, k) * tensor2.get(i, k);
+      }
+      get(*result) += *buffer1 * *buffer2;
+    }
+  }
+}
+}  // namespace detail
+
+template <int ConformalMatterScale>
+void add_hamiltonian_sources(
+    const gsl::not_null<Scalar<DataVector>*> hamiltonian_constraint,
+    const Scalar<DataVector>& conformal_energy_density,
+    const Scalar<DataVector>& extrinsic_curvature_trace,
+    const Scalar<DataVector>& conformal_factor) noexcept {
+  // See BaumgarteShapiro, Eq. (3.110)
+  get(*hamiltonian_constraint) +=
+      (square(get(extrinsic_curvature_trace)) / 12. -
+       2. * M_PI * get(conformal_energy_density) /
+           pow<ConformalMatterScale>(get(conformal_factor))) *
+      pow<5>(get(conformal_factor));
+}
+
+template <int ConformalMatterScale>
+void add_linearized_hamiltonian_sources(
+    const gsl::not_null<Scalar<DataVector>*> linearized_hamiltonian_constraint,
+    const Scalar<DataVector>& conformal_energy_density,
+    const Scalar<DataVector>& extrinsic_curvature_trace,
+    const Scalar<DataVector>& conformal_factor,
+    const Scalar<DataVector>& conformal_factor_correction) noexcept {
+  get(*linearized_hamiltonian_constraint) +=
+      ((5. / 12.) * square(get(extrinsic_curvature_trace)) -
+       2. * (5. - ConformalMatterScale) * M_PI * get(conformal_energy_density) /
+           pow<ConformalMatterScale>(get(conformal_factor))) *
+      pow<4>(get(conformal_factor)) * get(conformal_factor_correction);
+}
+
+void add_distortion_hamiltonian_sources(
+    const gsl::not_null<Scalar<DataVector>*> hamiltonian_constraint,
+    const Scalar<DataVector>&
+        longitudinal_shift_minus_dt_conformal_metric_over_lapse_square,
+    const Scalar<DataVector>& conformal_factor) noexcept {
+  // See BaumgarteShapiro, Eq. (3.110) and (3.112)
+  get(*hamiltonian_constraint) -=
+      0.03125 * pow<5>(get(conformal_factor)) *
```

Comment:
[optional] I have no objection to having numbers like `0.03125` here, but it would be nice to have a comment of what fraction it is :) If you decide to add a comment for this here, could you also do so in the below functions?

Reply:
done 👍